### PR TITLE
[feat] Add a transform for projects using workspaces

### PR DESCRIPTION
**Pull request by zhanwang626**
## Summary
Some projects using workspaces which may put their `babel-plugin-ember-test-data` config in a sub directory. In this case, we need to add `projectRoot: '../..'` in the config:
```js
'use strict';

const EmberApp = require('ember-cli/lib/broccoli/ember-app');

module.exports = function (defaults) {
  let app = new EmberApp(defaults, {
    babel: {
      plugins: [
        [
          require.resolve('babel-plugin-ember-test-metadata'),
          {
            enabled: !!process.env.BABEL_TEST_METADATA,
            packageName: defaults.project.pkg.name,
            isUsingEmbroider: !!process.env.EMBROIDER,
            packageRoot: '../..',
          },
        ],
      ],
    },
  });

  // additional configuration

  return app.toTree();
};
```
This PR also abstract some functions to `utils.js`
And added more test cases for the two transforms.

### CODE REVIEW

**Comment by zhanwang626** [comment]
Didn't figure out a good way to specify `projectRoot` other than hardcode.

File: transforms/utils.js [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        j.memberExpression(j.identifier('defaults'), j.identifier('project')),
+        j.identifier('pkg')
+      ),
+      j.identifier('name')
+    )
+  );
+}
+
+function getProjectRootProperty(j) {
+  return j.property('init', j.identifier('projectRoot'), j.literal('../..'));
```
Comment (zhanwang626) [inline transforms/utils.js:145]: Didn't figure out a good way to specify projectRoot other than hardcode.

**Comment by drewlee** [comment]
I think we're conflating two different ideas here as the same thing, in-repo addons/engines and workspaces. An app can have in-repo addons without using workspaces and vice-versa. In the former case, the `projectRoot` option wouldn't be needed.

For in-repo addons/engines, we would also need to inject the Babel config into each addon's `index.js` file, as well as install and configure [ember-add-in-repo-tests](https://github.com/ember-add-in-repo-tests/ember-add-in-repo-tests) if it doesn't already exist in `ember-cli-build.js`.

In summary, we'd need two different transforms. One for handling workspaces, the other for handling in-repo addons, as they are two different things.

**Comment by zhanwang626** [comment]
> I think we're conflating two different ideas here as the same thing, in-repo addons/engines and workspaces. An app can have in-repo addons without using workspaces and vice-versa. In the former case, the `projectRoot` option wouldn't be needed.
> 
> For in-repo addons/engines, we would also need to inject the Babel config into each addon's `index.js` file, as well as install and configure [ember-add-in-repo-tests](https://github.com/ember-add-in-repo-tests/ember-add-in-repo-tests) if it doesn't already exist in `ember-cli-build.js`.
> 
> In summary, we'd need two different transforms. One for handling workspaces, the other for handling in-repo addons, as they are two different things.

Okay, I see. Thanks for the explanation @drewlee ! I think this PR is more likely to handle the workspaces case. Will change the PR's description and diff to reflect it.

File: transforms/setup-config-with-workspaces/README.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+**Input** (<small>[basic.input.js](transforms/setup-config-with-workspaces/__testfixtures__/basic.input.js)</small>):
+```js
+'use strict';
+
+const EmberApp = require('ember-cli/lib/broccoli/ember-app');
+
+module.exports = function (defaults) {
+  let app = new EmberApp(defaults, {
+    'ember-cli-pemberly-i18n': i18nConfig,
```
Comment (drewlee) [inline transforms/setup-config-with-workspaces/README.md]: Can we genericize these examples so that we're not using internal LI specific tooling?

File: transforms/setup-config-with-workspaces/README.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+**Output** (<small>[basic.output.js](transforms/setup-config-with-workspaces/__testfixtures__/basic.output.js)</small>):
+```js
+'use strict';
+
+const EmberApp = require('ember-cli/lib/broccoli/ember-app');
+
+module.exports = function (defaults) {
+  let app = new EmberApp(defaults, {
+    'ember-cli-pemberly-i18n': i18nConfig,
```
Comment (drewlee) [inline transforms/setup-config-with-workspaces/README.md]: Can we genericize these examples so that we're not using internal LI specific tooling?

File: transforms/setup-config-with-workspaces/__testfixtures__/basic.input.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+'use strict';
+
+const EmberApp = require('ember-cli/lib/broccoli/ember-app');
+
+module.exports = function (defaults) {
+  let app = new EmberApp(defaults, {
+    'ember-cli-pemberly-i18n': i18nConfig,
```
Comment (drewlee) [inline transforms/setup-config-with-workspaces/__testfixtures__/basic.input.js]: Can we genericize these fixtures so that we're not using internal LI specific tooling?

File: transforms/setup-config-with-workspaces/__testfixtures__/basic.output.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
+'use strict';
+
+const EmberApp = require('ember-cli/lib/broccoli/ember-app');
+
+module.exports = function (defaults) {
+  let app = new EmberApp(defaults, {
+    'ember-cli-pemberly-i18n': i18nConfig,
```
Comment (drewlee) [inline transforms/setup-config-with-workspaces/__testfixtures__/basic.output.js]: Can we genericize these fixtures so that we're not using internal LI specific tooling?

File: transforms/setup-config-with-workspaces/__testfixtures__/basic.output.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+      includeModules: ['solid-gauge'],
+    },
+
+    babel: {
+      plugins: [[require.resolve('babel-plugin-ember-test-metadata'), {
+        enabled: !!process.env.BABEL_TEST_METADATA,
+        packageName: defaults.project.pkg.name,
+        isUsingEmbroider: !!process.env.EMBROIDER,
+        projectRoot: '../..',
+      }]],
```
Comment (drewlee) [inline transforms/setup-config-with-workspaces/__testfixtures__/basic.output.js]: Can we add support and test for the use case where the config exists but needs the `projectRoot` option added? For example, an app may get converted to use workspaces, but the owners never updated the config.

Before:
```js
plugins: [[require.resolve('babel-plugin-ember-test-metadata'), {
        enabled: !!process.env.BABEL_TEST_METADATA,
        packageName: defaults.project.pkg.name,
      }]],
```

After:
```js
plugins: [[require.resolve('babel-plugin-ember-test-metadata'), {
        enabled: !!process.env.BABEL_TEST_METADATA,
        packageName: defaults.project.pkg.name,
        projectRoot: '../..',
      }]],
```

WDYT?

File: transforms/setup-config-with-workspaces/README.md (hunk 2 above)
Comment (zhanwang626) [reply transforms/setup-config-with-workspaces/README.md]: Yeah, no problem

File: transforms/setup-config-with-workspaces/__testfixtures__/basic.output.js (hunk 6 above)
Comment (zhanwang626) [reply transforms/setup-config-with-workspaces/__testfixtures__/basic.output.js]: Yeah, no problem. I'm lazy to add it before because we have some similar tests for this in `setup-config` transform.

File: transforms/utils.js (hunk 1 above)
Comment (drewlee) [reply transforms/utils.js:145]: I think this is fine for 98% of all cases.

File: transforms/setup-config-with-workspaces/__testfixtures__/basic.input.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
+'use strict';
+
+const EmberApp = require('ember-cli/lib/broccoli/ember-app');
+
+module.exports = function (defaults) {
+  let app = new EmberApp(defaults, {
+    emberHighCharts: {
```
Comment (drewlee) [inline transforms/setup-config-with-workspaces/__testfixtures__/basic.input.js]: We should add a test case (and possibly handle, if not already done so) an app that already includes some Babel plugins in `ember-cli-build.js`, as most apps are set up this way.

For example:
Before
```js
  plugins: [
      '@babel/plugin-proposal-optional-chaining',
      '@babel/plugin-proposal-nullish-coalescing-operator',
      '@babel/plugin-proposal-object-rest-spread',
  ]
```
After
```js
  plugins: [
      '@babel/plugin-proposal-optional-chaining',
      '@babel/plugin-proposal-nullish-coalescing-operator',
      '@babel/plugin-proposal-object-rest-spread',
      [
        require.resolve('babel-plugin-ember-test-metadata'),
        {
          enabled: !!process.env.BABEL_TEST_METADATA,
          packageName: defaults.project.pkg.name,
          isUsingEmbroider: !!process.env.EMBROIDER,
          projectRoot: '../..',
        },
      ],
  ]
```

File: transforms/setup-config-with-workspaces/__testfixtures__/basic.input.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
+'use strict';
+
+const EmberApp = require('ember-cli/lib/broccoli/ember-app');
+
+module.exports = function (defaults) {
+  let app = new EmberApp(defaults, {
```
Comment (drewlee) [inline transforms/setup-config-with-workspaces/__testfixtures__/basic.input.js]: Some apps define their options in a variable instead of inline, would these transforms be able to handle this case?

For example:
```js
  const options = {
    emberHighCharts: {
      includeHighCharts: true,
      includeHighStock: false,
      includeHighChartsMore: true,
      includeHighCharts3D: true,
      includeModules: ['solid-gauge'],
    },
  };

  let app = new EmberApp(defaults, options);
  return app.toTree();
```

File: transforms/setup-config-with-workspaces/__testfixtures__/basic.input.js (hunk 7 above)
Comment (zhanwang626) [reply transforms/setup-config-with-workspaces/__testfixtures__/basic.input.js]: Yeah, this case has been already handled in another transform's test. But I can add it to this transform as well.

File: transforms/setup-config-with-workspaces/__testfixtures__/basic.input.js (hunk 8 above)
Comment (zhanwang626) [reply transforms/setup-config-with-workspaces/__testfixtures__/basic.input.js]: Added a test case for this.

File: transforms/setup-config-with-workspaces/__testfixtures__/basic.input.js (hunk 7 above)
Comment (zhanwang626) [reply transforms/setup-config-with-workspaces/__testfixtures__/basic.input.js]: Added test cases.

File: transforms/utils.js [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+    })
+    .forEach((path) => {
+      optionsName = path.node.arguments[1].name;
+    });
+
+  if (optionsName === undefined) {
+    configurationType = CONFIGURATION_TYPE.INLINE;
+  } else {
+    configurationType = CONFIGURATION_TYPE.USE_OPTIONS;
+  }
```
Comment (drewlee) [inline transforms/utils.js]: Suggestion for brevity:
```js
configurationType = CONFIGURATION_TYPE[optionsName === undefined ? 'INLINE' : 'USE_OPTIONS'];
```

**Review by drewlee — APPROVED** [review]
Minor suggestion, you can take it or leave it. Otherwise, LGTM.
